### PR TITLE
chore: Refactoring of cancellation handling in REGAPIC

### PR DESCRIPTION
**Pull request by jskeet**
This doesn't fix either #652 or #655, but is a stepping stone to improving both.

### CODE REVIEW

File: Google.Api.Gax.Grpc/Rest/RpcCancellationContext.cs [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        string rpcName,
+        CancellationTokenSource deadlineCancellationTokenSource,
+        CancellationToken callCancellationToken) =>
+        new RpcCancellationContext(rpcName, deadlineCancellationTokenSource, callCancellationToken);
+
+    /// <summary>
+    /// Creates a cancellation context from the given call options.
+    /// </summary>
+    /// <param name="rpcName">The name of the RPC, used to report exceptions.</param>
+    /// <param name="options"></param>
```
Comment (amanda-tarafa) [inline Google.Api.Gax.Grpc/Rest/RpcCancellationContext.cs:58]: nit: Empty param description.

File: Google.Api.Gax.Grpc/Rest/RpcCancellationContext.cs [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    /// Combines <paramref name="cancellationToken"/> with the RPC's cancellation tokens, and waits for
+    /// the task provided by <paramref name="taskProvider"/> to complete, converting any <see cref="OperationCanceledException"/>
+    /// into an <see cref="RpcException"/> with an appropriate status (DeadlineExceeded or Cancelled depending
+    /// on which token was responsible for the original exception).
+    /// </summary>
+    /// <param name="taskProvider">A function which returns a task to await.</param>
+    /// <param name="cancellationToken">An additional cancellation token, defaulting to "no extra cancellation token,
+    /// just use the RPC's cancellation tokens".</param>
+    /// <returns>A task which will complete when the provided task completes.</returns>
+    public async Task AwaitAsync(Func<CancellationToken, Task> taskProvider, CancellationToken cancellationToken = default)
```
Comment (amanda-tarafa) [inline Google.Api.Gax.Grpc/Rest/RpcCancellationContext.cs:85]: maybe the name of this method should be `RunAsync` so that callers do `await RunAsync` and not `await AwaitAsync`

File: Google.Api.Gax.Grpc/Rest/RpcCancellationContext.cs [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    /// the task provided by <paramref name="taskProvider"/> to complete, converting any <see cref="OperationCanceledException"/>
+    /// into an <see cref="RpcException"/> with an appropriate status (DeadlineExceeded or Cancelled depending
+    /// on which token was responsible for the original exception).
+    /// </summary>
+    /// <typeparam name="T">The type of the task provided by <paramref name="taskProvider"/>.</typeparam>
+    /// <param name="taskProvider">A function which returns a task to await.</param>
+    /// <param name="cancellationToken">An additional cancellation token, defaulting to "no extra cancellation token,
+    /// just use the RPC's cancellation tokens".</param>
+    /// <returns>A task which will complete when the provided task completes.</returns>
+    public async Task<T> AwaitAsync<T>(Func<CancellationToken, Task<T>> taskProvider, CancellationToken cancellationToken = default)
```
Comment (amanda-tarafa) [inline Google.Api.Gax.Grpc/Rest/RpcCancellationContext.cs:111]: And same for this one AwaitAsync to RunAsync.

File: Google.Api.Gax.Grpc.Tests/Rest/RpcCancellationContextTest.cs [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+using Grpc.Core;
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace Google.Api.Gax.Grpc.Rest.Tests;
+
+public class RpcCancellationContextTest
```
Comment (amanda-tarafa) [inline Google.Api.Gax.Grpc.Tests/Rest/RpcCancellationContextTest.cs:16]: Maybe add one or two tests that receive a null deadline cancellation token source, mostly for documentation and so that we don't inadvertenly change that behaviour later.

File: Google.Api.Gax.Grpc/Rest/RestChannel.cs [hunk 5]
```diff
@@ -106,22 +101,8 @@ public RestChannel(RestServiceCollection serviceCollection, string endpoint, Cha
...
-                throw new RpcException(new Status(StatusCode.DeadlineExceeded, $"RPC '{restMethod.FullName}' timed out", ex));
-            }
-            catch (OperationCanceledException ex)
-            {
-                throw new RpcException(new Status(StatusCode.Cancelled, $"RPC '{restMethod.FullName}' was cancelled", ex));
-            }
-
-            return httpResponseMessage;
+            await cancellationContext.AwaitAsync(token => AddAuthHeadersAsync(httpRequest, restMethod, token)).ConfigureAwait(false);
+            return await cancellationContext.AwaitAsync(token => _httpClient.SendAsync(httpRequest, httpCompletionOption, token)).ConfigureAwait(false);
```
Comment (amanda-tarafa) [inline Google.Api.Gax.Grpc/Rest/RestChannel.cs:105]: Maybe this is more readable? Not sure though.

```suggestion
            return await cancellationContext.AwaitAsync(token =>
            {
                await AddAuthHeadersAsync(httpRequest, restMethod, token).ConfigureAwait(false);
                return await _httpClient.SendAsync(httpRequest, httpCompletionOption, token).ConfigureAwait(false);
            }).ConfigureAwait(false);
```

**Review by amanda-tarafa — APPROVED** [review]
LGTM, yep, this is way neater.

File: Google.Api.Gax.Grpc/Rest/RpcCancellationContext.cs (hunk 1 above)
Comment (jskeet) [reply Google.Api.Gax.Grpc/Rest/RpcCancellationContext.cs:58]: Done.

File: Google.Api.Gax.Grpc/Rest/RpcCancellationContext.cs (hunk 2 above)
Comment (jskeet) [reply Google.Api.Gax.Grpc/Rest/RpcCancellationContext.cs:85]: Yes, good plan. Will do.

File: Google.Api.Gax.Grpc/Rest/RestChannel.cs (hunk 5 above)
Comment (jskeet) [reply Google.Api.Gax.Grpc/Rest/RestChannel.cs:105]: Will try a local method - I think that'll be cleaner than either.

File: Google.Api.Gax.Grpc.Tests/Rest/RpcCancellationContextTest.cs (hunk 4 above)
Comment (jskeet) [reply Google.Api.Gax.Grpc.Tests/Rest/RpcCancellationContextTest.cs:16]: Done.

**Review by amanda-tarafa — APPROVED** [review]
LGTM
